general-superstaq/general_superstaq/_models.py

```diff
@@ -0,0 +1,291 @@
+    AWAITING_SUBMISSION = "awaiting_submission"
+    """The job is waiting for a worker to submit the circuit to an external device."""
+    AWAITING_SIMULATION = "awaiting_simulation"
+    """The job is waiting for a worker to simulate."""
+    AWAITING_CONVERSION = "awaiting_conversion"
+    """The job is waiting for a worker to convert."""
+    RUNNING = "running"
+    """The job is being run by a worker."""
+    COMPLETED = "completed"
+    """The job is completed."""
+    FAILED = "failed"
+    """The job failed. A reason should be stored in the job."""
+    CANCELLED = "cancelled"
+    """The job was cancelled. A reason should be stored in the job."""
+    UNRECOGNIZED = "unrecognized"
+    """Something has gone wrong! (Treated as terminal)"""
+    PENDING = "pending"
+    """When a job has been submitted to an external provider but that provider has
+    not yet run the job."""
+    DELETED = "deleted"
+    """When a job has been deleted."""
+
+
+TERMINAL_CIRCUIT_STATES = [
+    CircuitStatus.COMPLETED,
+    CircuitStatus.FAILED,
+    CircuitStatus.CANCELLED,
+    CircuitStatus.UNRECOGNIZED,
+    CircuitStatus.DELETED,
+]
+
+UNSUCCESSFUL_CIRCUIT_STATES = [
+    CircuitStatus.CANCELLED,
+    CircuitStatus.FAILED,
+    CircuitStatus.DELETED,
+]
+
+
+class DefaultPydanticModel(pydantic.BaseModel, use_enum_values=True, extra="forbid"):
+    """Default pydantic model used across the superstaq server."""
+
+
+class JobData(DefaultPydanticModel):
+    """A class to store data for a Superstaq job which is returned through to the client."""
+
+    job_type: JobType
+    statuses: list[CircuitStatus]
+    status_messages: list[str | None]
+    user_email: pydantic.EmailStr
+    target: str
+    provider_id: list[str | None]
+    num_circuits: int
+    compiled_circuit_type: CircuitType
+    compiled_circuits: list[str | None]
+    input_circuits: list[str]
+    input_circuit_type: CircuitType
+    pulse_gate_circuits: list[str | None]
+    counts: list[dict[str, int] | None]
+    state_vectors: list[str | None]
+    results_dicts: list[str | None]
+    num_qubits: list[int]
+    shots: list[int]
+    dry_run: bool
+    submission_timestamp: datetime.datetime
+    last_updated_timestamp: list[datetime.datetime | None]
+    initial_logical_to_physicals: list[str | None]
+    final_logical_to_physicals: list[str | None]
+
+
+class NewJob(DefaultPydanticModel):
+    """The data model for submitting new jobs."""
+
+    job_type: JobType
+    target: str
+    circuits: str
+    circuit_type: CircuitType
+    compiled_circuits: str | None = pydantic.Field(default=None)
+    compiled_circuit_type: CircuitType | None = pydantic.Field(default=None)
+    shots: int = pydantic.Field(default=0, ge=0)
+    dry_run: bool = pydantic.Field(default=False)
+    sim_method: SimMethod | None = pydantic.Field(default=None, validate_default=True)
+    priority: int = pydantic.Field(default=0)
+    options_dict: dict[str, Any] = pydantic.Field(default={})
```

Comment:
this might be a good time to think about retiring the options dict. it was originally mostly intended as a stopgap so that we could make things flexible as we refined our interfaces, now that we're defining these models we should maybe to start breaking things into their own fields/models (and maybe just use for `extra="allow"` for experimental/wip/non-publicized fields?)

(not sure how we want to organize things though - we could put everything in the top level model, or keep a single `options` field w/ a model for possible arguments, or start separating e.g. compilation options/user credentials/simulation parameters into their own models...

Reply:
This is top of my to-do list once the initial server is up and running. As per my comment above about extra="forbid", the client side models should be fully enumerated.

I imagine that the best approach will be to have different models for different types of jobs etc., set up in some hierarchical way this will ensure that everything gets validated automatically by the server. This is part of the job refactor to-do list

For now I suggest we keep the `options_dict` field to maintain compatibility with existing Superstaq method.

Reply:
yep totally agree, just wanted to flag it in case it had any implications for how we go about organizing things